protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -75,7 +79,41 @@ public synchronized void clear() {
     }
 
     public synchronized void addUid(MessageUid uid) {
-        this.addMapping(nextMsn(), uid);
+        if (isLastUid(uid)) {
+            addMapping(nextMsn(), uid);
+        } else {
+            addUidInMiddle(uid);
+        }
+    }
+
+    private boolean isLastUid(MessageUid uid) {
+        return msnToUid.isEmpty() ||
+            uid.asLong() > msnToUid.get(getLastMsn()).asLong();
+    }
+
+    private boolean alreadyContains(MessageUid uid) {
+        return msnToUid.containsValue(uid);
+    }
+
+    private void addUidInMiddle(MessageUid uid) {
+        List<MessageUid> aboveUids = removeAndGetAboveUidSortedInIncreasingOrder(uid);
+        addMapping(nextMsn(), uid);
+        for (MessageUid aboveUid : aboveUids) {
+            addMapping(nextMsn(), aboveUid);
+        }
+    }
+
+    private List<MessageUid> removeAndGetAboveUidSortedInIncreasingOrder(MessageUid uid) {
+        ImmutableList.Builder<MessageUid> result = ImmutableList.builder();
+        int position = getLastMsn();
+        Optional<MessageUid> maxUid = getUid(position);
+        while (maxUid.isPresent() && uid.asLong() < maxUid.get().asLong()) {
```

Comment:
**indexOf** operation is in O(n) in ArrayList, which makes me thinks it's not a good idea.